src/lib/Repository/Validator/TargetContentValidator.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Core\Repository\Validator;
+
+use eZ\Publish\API\Repository\ContentService;
+use eZ\Publish\API\Repository\ContentTypeService;
+use eZ\Publish\API\Repository\Exceptions\NotFoundException;
+use eZ\Publish\API\Repository\Exceptions\UnauthorizedException;
+use eZ\Publish\Core\FieldType\ValidationError;
+
+/**
+ * Validator for checking existence of content and its content type.
+ */
+final class TargetContentValidator implements TargetContentValidatorInterface
+{
+    /** @var \eZ\Publish\API\Repository\ContentService */
+    private $contentService;
+
+    /** @var \eZ\Publish\API\Repository\ContentTypeService */
+    private $contentTypeService;
+
+    public function __construct(
+        ContentService $contentService,
+        ContentTypeService $contentTypeService
+    ) {
+        $this->contentService = $contentService;
+        $this->contentTypeService = $contentTypeService;
+    }
+
+    public function validate(int $value, array $allowedContentTypes = []): ?ValidationError
+    {
+        try {
+            $contentInfo = $this->contentService->loadContentInfo($value);
+            $contentType = $this->contentTypeService->loadContentType($contentInfo->contentTypeId);
+
+            if (!empty($allowedContentTypes) && !in_array($contentType->identifier, $allowedContentTypes, true)) {
+                return new ValidationError(
+                    'Content Type %contentTypeIdentifier% is not a valid relation target',
+                    null,
+                    [
+                        '%contentTypeIdentifier%' => $contentType->identifier,
+                    ],
+                    'targetContentId'
+                );
+            }
+        } catch (NotFoundException | UnauthorizedException $e) {
```

Comment:
Maybe check should be done using `sudo` or on SPI layer?

Reply:
I don't think that is necessary. Being incapable of reaching a certain Content would mean that it is not a valid relationship target, correct? And we don't really need to treat them separately, unless we accept that user could be informed about his inability to reach a certain object (which in essence let's him know that it exists?).

Reply:
> Maybe check should be done using `sudo` or on SPI layer? (open question)

Depends on the expected validation logic - a principle. If inability to access `ContentInfo` and/or `ContentType` should validate as an error, then this should stay as is. If we intend to check for the existence of Content Types only and this catch is just an attempt to "fix holes", then indeed Persistence layer[1] should be injected instead. This is the question for @barw4 I guess.

[1] SPI expands to Service Provider Interface and does not fit really this context, we should stop using this misleading terminology, especially that it disappeared in Ibexa 4.x. It's Persistence Layer :upside_down_face:

Reply:
@alongosz I agree with @Steveb-p, we shouldn't validate such content if a user doesn't have access to it - this won't interfere in the UI. When it comes to the API one can always change the user context if he wants to validate content as sudo.